Like feature and page ordering features

A5rocks opened the pull request:
According to sublime merge, originally `like-feature-(and-page-ranking)` was branched from `add-page-render-functionality`, so maybe we should pull those changes back in? This should lead to having to review less lines!

Replaces https://github.com/Artemis21/summer-code-jam-2020/pull/26

tameTNT committed:
Add Like model to models.py

Add new model to allow for like functionality where a user can like a
webpage to improve its ordering on the 'best pages' page.

tameTNT committed:
Add Like model to models.py

Add new model to allow for like functionality where a user can like a
webpage to improve its ordering on the 'best pages' page.

tameTNT committed:
Add like count/popularity column to 'best pages'

Add a new column to the webpage_list.html template which tracks the
total number of likes a Webpage has received from users.

tameTNT committed:
Add like count/popularity column to 'best pages'

Add a new column to the webpage_list.html template which tracks the
total number of likes a Webpage has received from users.

tameTNT committed:
Finish Like model and page ordering

Register Like model with admin.py.
Remove deprecated votes attribute from Webpage model - this is now
handled by the fk in the Like model.
Implement ordering in WebpageListView, so webpages are now sorted by
number of likes on the 'Best Pages' ('/pages/') page.

tameTNT committed:
Finish Like model and page ordering

Register Like model with admin.py.
Remove deprecated votes attribute from Webpage model - this is now
handled by the fk in the Like model.
Implement ordering in WebpageListView, so webpages are now sorted by
number of likes on the 'Best Pages' ('/pages/') page.

tameTNT committed:
Merge remote-tracking branch 'origin/like-feature-(and-page-ranking)' into like-feature-(and-page-ranking)

tameTNT committed:
Add more num likes labels and enable pagination

Edit urls.py to conform with no leading '/' convention.
Actually implement pagination code on webpage_list.html with buttons to
go to the next/previous page etc.

tameTNT committed:
Improve basic page functionality

Add '?next={{ request.path }}' to end of urls to ensure user is
redirected back to the page where they clicked the button. More of a
convenience feature than anything!

Added clarification to user_detail.html so that the page says if the
user hasn't created any pages.

Corrected typo in register.html.

tameTNT committed:
Fully add 'like' feature to relevant pages

Users can now, if logged in, like/unlike pages (but not their own!)
A webpage's like count is now displayed in various locations - including
on the 'best pages' page and the page itself (webpage_detail.html).

Minor sorting improvements in UserDetailView to sort by edit datetime.

tameTNT committed:
Add a 'recent pages' page

New url for recent pages, 'pages-recent/', and by extension for 'best'
also,'pages-top/'. These are both WebpageListView views, just with
different ordering values.
Update main.html to show this new navigation option.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/base.html:
```diff
@@ -35,10 +35,10 @@
     {% if user.is_authenticated %}
         <a href="{% url 'user-detail' pk=user.id %}"> {{ user.username }}</a>
-        <a href="{% url 'logout' %}">Logout</a>
+        <a href="{% url 'logout' %}?next={{ request.path }}">Logout</a>
     {% else %}
-        <a href="{% url 'login' %}">Login</a>
-        <a href="{% url 'register' %}">Register</a>
+        <a href="{% url 'login' %}?next={{ request.path }}">Login</a>
+        <a href="{% url 'register' %}?next={{ request.path }}">Register</a>
     {% endif %}
```
We shouldn't indent in if blocks.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_detail.html:
```diff
@@ -2,6 +2,13 @@
 <div><h1>{{ webpage.user_title }}</h1></div>
+<div>
+    {{ num_likes }} like{{ num_likes|pluralize }}
+    <form method="POST" action="">
```
Isn't this blank `action` a hardcoded url?

A5rocks replied:
nevermind it doesn't seem to be?

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html:
```diff
@@ -17,12 +18,35 @@ <h2>Top pages</h2>
+<br>
+<div class="pagination">
+    <span class="step-links">
+        {% if page_obj.has_previous %}
+            <a href="?page=1">First page</a>
+            <a href="?page={{ page_obj.previous_page_number }}">Previous page</a>
+        {% endif %}
```
We shouldn't be indenting here

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_list.html:
```diff
@@ -17,12 +18,35 @@ <h2>Top pages</h2>
+            <a href="?page={{ page_obj.previous_page_number }}">Previous page</a>
+        {% endif %}
+
+        {% if page_obj.has_next %}
+            <a href="?page={{ page_obj.next_page_number }}">Next page</a>
+            <a href="?page={{ page_obj.paginator.num_pages }}">Last page</a>
+        {% endif %}
```
We shouldn't be indenting here

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/user_detail.html:
```diff
@@ -6,25 +6,36 @@ <h2>User {{ viewed_user.username }}</h2>
         <td>
             {% if "placeholder_img" not in webpage.thumbnail.url %}
-            <img src="{{ webpage.thumbnail.url }}">
+                <img src="{{ webpage.thumbnail.url }}">
             {% else %}
-            Go to page itself to generate a thumbnail for it.
+                Go to page itself to generate a thumbnail for it.
             {% endif %}
```
We shouldn't be indenting in an if block.

A5rocks commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -20,8 +20,11 @@ <h1><span>{{ webpage.user_title }}</span></h1>
+
 <p>
-    <a href="{% url 'webpage-detail' pagename=webpage.name %}">Comment on my page</a>
+    {{ num_likes }} like{{ num_likes|pluralize }}
+<br>
```
```suggestion
    {{ num_likes }} like{{ num_likes|pluralize }}<br>
```

A5rocks commented on jolly-jellyfish/src/page_maker/templates/registration/register.html:
```diff
@@ -1,7 +1,7 @@
-       <h2>Sing Up</h2>
+       <h2>Sign Up</h2>
        <form method="POST">
            {% csrf_token %}
```
This should probably be moved down an indentation level

A5rocks commented on jolly-jellyfish/src/page_maker/urls.py:
```diff
@@ -20,7 +20,7 @@
 )
 
 urlpatterns = [
```
These shouldn't end with `/`.

A5rocks commented on jolly-jellyfish/src/page_maker/views.py:
```diff
@@ -107,18 +116,44 @@ def get_context_data(self, **kwargs):
+
+        if request.user.is_authenticated:  # user is logged in
+            if request.user == self.object.author:  # it's the user's own page
+                context['like_response'] = "You can't like your own page!"
+            else:
+                like_obj, created = Like.objects.get_or_create(author=request.user, parent_page=self.object)
+                if created:  # this is first time the user has liked the webpage
+                    context['like_response'] = "Thanks for the like!"
+                else:
+                    like_obj.delete()
+                    context['like_response'] = "You've now unliked this page."
+        else:
+            context['like_response'] = "Please login to leave a like."
```
Double-quotes -> single-quotes.

A5rocks reviewed (commented):
A couple changes... (I can't "request changes" cause I made this PR :( )

A5rocks committed:
Merge branch 'add-page-render-functionality' into like-feature-(and-page-ranking)

A5rocks committed:
Make code more consistent

- Change HTML from 4-space indentation to 2-space indentation
- No indents in if blocks
- Prefer single-quotes unless the string has a contraction.

tameTNT reviewed (approved):
Looks good to me - but then I wrote the stuff so I would say that 😆

tameTNT commented on jolly-jellyfish/src/page_maker/templates/page_maker/webpage_view.html:
```diff
@@ -22,9 +22,8 @@ <h1><span>{{ webpage.user_title }}</span></h1>
-<br>
-    <a href="{% url 'webpage-detail' pagename=webpage.name %}">Like & comment on my page!</a>
+  {{ num_likes }} like{{ num_likes|pluralize }}<br>
+  <a href="{% url 'webpage-detail' pagename=webpage.name %}">Comment on my page</a>
```
```suggestion
  <a href="{% url 'webpage-detail' pagename=webpage.name %}">Like and comment on my page!</a>
```

tameTNT committed:
Change to Like and comment on my page

This additional text seems to have bee removed during one of the commit processes.

A5rocks committed:
Merge branch 'add-page-render-functionality' into like-feature-(and-page-ranking)

Artemis21 reviewed (approved):
LGTM
